backend/canisters/notifications_index/impl/src/model/fcm_token_store.rs

```diff
@@ -0,0 +1,133 @@
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+use types::UserId;
+
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord)]
+pub struct FcmToken(String);
+
+impl From<String> for FcmToken {
+    fn from(token: String) -> Self {
+        FcmToken(token)
+    }
+}
+impl From<FcmToken> for String {
+    fn from(token: FcmToken) -> Self {
+        token.0
+    }
+}
+impl AsRef<str> for FcmToken {
+    fn as_ref(&self) -> &str {
+        &self.0
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct FcmTokenStore {
+    // User tokens are stored in a BTreeSet to ensure uniqueness!
+    fcm_user_tokens: BTreeSet<(UserId, FcmToken)>,
+}
+
+impl FcmTokenStore {
+    /// Checks if a given FCM token exists!
+    pub fn check_token_exists(&self, token: &FcmToken) -> bool {
+        self.fcm_user_tokens.iter().any(|(_, t)| t == token)
+    }
+
+    /// Add a new FCM token, and assign it to a user.
+    ///
+    /// If the token already exists, it will not be added again - this also
+    /// ensures that the user does not have duplicate tokens, or that the same
+    /// token is not assigned to multiple users.
+    pub fn add_token(&mut self, token: FcmToken, user_id: UserId) -> Result<(), String> {
+        if !self.check_token_exists(&token) {
+            self.fcm_user_tokens.insert((user_id, token.clone()));
+            Ok(())
+        } else {
+            Err("Token already exists".to_string())
+        }
+    }
+
+    /// Remove a token from the store. We remove tokens when the Firebase
+    /// request to push notification with the token fails. This means that
+    /// the token is no longer valid or the user has uninstalled the app.
+    ///
+    /// If the user and token combo exists, it will remove the mapping, and we
+    /// confirm that with the OK result. If the token is not associated
+    /// with the user, remove operation returns false, and we return an error.
+    pub fn remove_token(&mut self, user_id: &UserId, token: &FcmToken) -> Result<(), String> {
+        if self.fcm_user_tokens.remove(&(*user_id, token.clone())) {
+            Ok(())
+        } else {
+            Err("Token is not associated with current user".to_string())
+        }
+    }
+
+    /// Get all FCM tokens associated with specific users! It is very likely
+    /// we'll have to push notifications to multiple users at once, so this
+    /// method allows us to retrieve all tokens for a set of users.
+    pub fn get_tokens_for_users(&self, user_ids: Vec<&UserId>) -> Vec<&FcmToken> {
```

Comment:
Probably better to make this `get_tokens_for_user`, since that is a fast lookup by using `self.fcm_user_tokens.range((user_id, String::default())..).take_while(|(u, _)| u == user_id)`